### DISCUSSION TITLE
MRF Dictionary Simulation

### DISCUSSION THREAD

**Discussion by nathanjbeaumont** [MRsources/MRzero-Core#92] [Q&A]
Hello,
I work with magnetic resonance fingerprinting (MRF) and I have been using EPG simulation in Matlab to generate my dictionary as was done in [this paper](https://pubmed.ncbi.nlm.nih.gov/25491018/). I would also like to have dictionary simulation working in MR0. The pulse sequence has 1000 RF pulses.

I created a phantom one voxel in size in MR0 and ran my pulse sequence on my GPU but it took a full 1 minute 3s to run `execute_graph` . Running EPG in Matlab for a single voxel takes 0.1 seconds. Does anyone have suggestions for how to simulate a dictionary like this?

**Comment by mzaiss** [comment]
Thank you for your question!

Well there are several answers to that question.

Is your fingerprinting sequence using encoding gradients? or do you just run the tr fa part?

with gradients the underlying pdg has to describe many more dephased states, that are not simulated at all in a epg.  also epg works on a fixed grid given by your timing, this is estimated in pdg and might not work as efficiently as on an actual fixed grid. and resulting number of states can still grow exponentially.

also the prepass graph might be large and considerable time needed for pruning the PDG graph.

lastly, we never optimized it comparing to a single epg, 
which is maybe something we want to do. maybe its just not efficiently programmed for a single voxel, as many 3D tensors are obsolete.

still, you can try to simulate a full image and see how much longer it takes. i guess 1 voxel or 600 voxels takes almost the same time. you could then store the actual magnetization state and generate your dictionary in parallel, then you maybe outperform EPG.


Still EPG is a wonderful tool, why exactly do you wanna switch for a single voxel?

best,
moritz

**Reply by nathanjbeaumont** [reply]
> still, you can try to simulate a full image and see how much longer it takes. i guess 1 voxel or 600 voxels takes almost the same time. you could then store the actual magnetization state and generate your dictionary in parallel, then you maybe outperform EPG.

So create a 600 voxel phantom? How would I store the actual magnetization state? `execute_graph` outputs a 1D signal.

Here is the code I have for 1 voxel simulation:
[mrZeroCoreClean.zip](https://github.com/user-attachments/files/18017886/mrZeroCoreClean.zip)

It has 1000 TRs with 1000 flip angles:
![image](https://github.com/user-attachments/assets/d40f6555-be11-459a-aafc-5b34d3305a0c)

No x or y gradients are applied during the dictionary simulation. There is one ADC read after each flip.

**Comment by mzaiss** [comment]
@SWeinmueller, can you show us how to get mag back?

**Reply by nathanjbeaumont** [reply]
Update: I realized you do need to have the gradients applied during dict simulation. Updated code is here:
https://drive.google.com/file/d/1VXjT_gmC7fVcc13yTvox8a1HEp-U7dj9/view?usp=sharing

You can see the single voxel dictionary signal matches fairly well to the simulated brain acquisition
![image](https://github.com/user-attachments/assets/54ef5798-8c20-435a-aaa6-70eb034debc3)

**Comment by SWeinmueller** [comment]
Example for FLASH: 

```
signal, mag_adc  = mr0.execute_graph(graph, seq, obj_p, **return_mag_adc=True**)

img = torch.zeros([*sz], dtype=torch.complex64)
mag = torch.stack([m[0] for m in mag_adc])
rows = 8
cols = 8
factor = int(sz[0] / rows)

# Generate plots for each subplot
plt.figure(dpi=300)
plt.suptitle("Transversal adc Mag", fontsize=16)
for i in range(rows):
    for j in range(cols):
        plt.subplot(rows, cols, i * cols + j + 1)
        img[obj_p.recover().PD>0] = mag[i*factor,j*factor,:]
        util.plot3D(img.abs())
        plt.clim(0,0.15)
        plt.axis('off')
# Show the plot
plt.show()

# Generate plots for each subplot
plt.figure(dpi=300)
plt.suptitle("Transversal adc Mag", fontsize=16)
for i in range(rows):
    for j in range(cols):
        plt.subplot(rows, cols, i * cols + j + 1)
        img[obj_p.recover().PD>0] = mag[i*factor,j*factor,:]
        util.plot3D(img.angle())
        plt.clim(0,0.15)
        plt.axis('off')
# Show the plot
plt.show()
```

![grafik](https://github.com/user-attachments/assets/3973b402-b1c4-4fc8-8fe8-11cfeeb390e3)

![grafik](https://github.com/user-attachments/assets/a75298b9-a367-448d-bc24-8012bc1314f5)

**Reply by SWeinmueller** [reply]
You can add return_mag_adc=True to execute graph, that returns a list of signals with length repetitions, e.g. 128 for resolution 128. Then, there is the number of paths, number of adc_points, and then number of voxels, which are simulated.

Using obj_p.recover() you can geht back the phantom and can check where obj_p.recover().PD>0.

You get back how the image would looks like if you could acquire at a certain position in k-space the complete image at ones.
